mac/keg_relocate: use relative install names

gerlero opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [X] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew typecheck` with your changes locally?
- [X] Have you successfully run `brew tests` with your changes locally?

-----

Adds to the logic for how `install_name`s and `rpath`s are fixed (after a build) and relocated (from bottles) on macOS, so that the installed binaries will load other Homebrew-installed dynamic libraries using relative paths instead of absolute paths. 

For example, a reference to a library with an `install_name` currently set in an executable as:

 `/opt/homebrew/Cellar/<same_formula>/<same_version>/lib/lib<some_lib>.dylib` 

will be changed to something like:

 `@loader_path/../lib/lib<some_lib>.dylib` 

instead, and references to libraries from dependencies are changed to:

`@loader_path/../../../../opt/<other_formula>/lib/lib<other_lib>.dylib`.

It should cover both bottled installs as well as builds from source. I've tried both—with default and custom Homebrew prefixes—and didn't run into any issues.

**Note**: there are also other ways to accomplish this same behavior (e.g., by adding a relative `rpath` entry to the binaries pointing to some known anchor point—such as the Homebrew prefix—, and then rewriting all install names to be prefixed with `@rpath`), but this ways keeps the code changes small. Also, Homebrew will already add a `@loader_path`-relative path to fix some unqualified library entries, while some formulae already set `rpath`s for their own purposes (which we shouldn't interfere with).

### Rationale

The point of this is proposed change is to have a Homebrew install be able to survive a filesystem move of the entire prefix, as long as the internal directory structure is preserved (not all formulae can gracefully handle being moved right now, but future PRs at Homebrew/core could help with that). It is a small step towards the Homebrew installation itself eventually being fully relocatable/"portable", which for instance would help with some currently non-standard uses of Homebrew, such as project-specific installs or as a bundler for application dependencies.

I understand that anything other than installing at the default prefix is outside the scope of official support, which is why I went for a PR instead of opening an issue; and why I've intentionally kept it as simple as I could. I hope this change can be considered for review.

EDIT: fix example path

MikeMcQuaid reviewed (commented):
Seems reasonable to me so far but I'll want a few others to give it some 👀.

This seems like a big enough change that it may warrant being optionally enabled by an environment variable first in CI, then for HOMEBREW_DEVELOPERs, etc. so we don't merge a single PR that potentially breaks a lot of people's setups.

Would love to hear how you've tested this and what e.g formulae/systems/prefixes you tried it with.

> It is a small step towards the Homebrew installation itself eventually being fully relocatable/"portable", which for instance would help with some currently non-standard uses of Homebrew, such as project-specific installs or as a bundler for application dependencies.

This is desirable, I agree, but the main issue here I'd argue is not lack of relocatability of Homebrew's prefix but the lack of relocatability of many of Homebrew's bottles. That's not to block this PR, if/when it works we should consider merging it, but just to note that if you want to tackle this problem then something like https://github.com/Homebrew/brew/issues/10846 would probably be the highest impact next direction after this PR.

Thanks for the PR!

carlocab commented:
To make sure we're on the same page about all the consequences of this pull request: can you show an example of the output of `otool -L` for a dylib before and after this change?

gerlero commented:
> Seems reasonable to me so far but I'll want a few others to give it some 👀.
> 
> This seems like a big enough change that it may warrant being optionally enabled by an environment variable first in CI, then for HOMEBREW_DEVELOPERs, etc. so we don't merge a single PR that potentially breaks a lot of people's setups.

Agree completely. This needs many careful eyes and ideally should take effect gradually, as things can easily go ugly if we happen to miss something.

> Would love to hear how you've tested this and what e.g formulae/systems/prefixes you tried it with.

I've tested mostly by building [this project of mine](https://github.com/gerlero/openfoam-app), which by default uses Homebrew to bundle dependencies (custom Homebrew prefix at `/Volumes/<something>/usr` in a case-sensitive FS), but also has the option to use a system Homebrew install (a the default prefix); so I tried both. Here's the [Brewfile](https://github.com/gerlero/openfoam-app/blob/main/Brewfile) with the formulae that that project uses; it uses a custom tap, and in the case of the custom prefix many of the formulae need to be built from source.

I've also tried requesting a build from source in the default prefix, bottling it, and then installing that bottle to see if the relative paths remained there untouched.

I've then tried moving the entire Homebrew prefix. In this case for example, [`open-mpi`](https://formulae.brew.sh/formula/open-mpi) will still fail to work properly. However, as expected, the error changes from the loader not finding the required libraries, to the package itself not finding its own stuff (i.e., the other reason you mention that affects relocatability). In this particular case, [upstream offers instructions for making the install relocatable](https://www.open-mpi.org/faq/?category=building#installdirs), but I haven't gotten it to work yet.

This is by no means thorough, but it does cover a bunch of different cases. A major thing is that I've only tested Apple silicon (and not Intel). Does the CI also test formulae installed from bottles and from source? (all this time I'm assuming it does; otherwise we'd need to be even more careful with this...)

> This is desirable, I agree, but the main issue here I'd argue is not lack of relocatability of Homebrew's prefix but the lack of relocatability of many of Homebrew's bottles. That's not to block this PR, if/when it works we should consider merging it, but just to note that if you want to tackle this problem then something like #10846 would probably be the highest impact next direction after this PR.

Yes. This is another problem, as `open-mpi` shows. Thanks for the link, I've not though about directly patching the binary contents before, so it kind of scares me (I'd rather try to add relocatability on a formula-by-formula basis); but if Conda does it successfully then it can't be _that_ bad (and would be more resilient against updates to formulae that unexpectedly break relocatability). Configuration and script files can also contain hardcode paths, so I guess that would need addressing too.

From the changes I've made to `mac/keg_relocate.rb`, I also saw that it some instances Homebrew itself will set an absolute path to its Perl or Java installs (which should be made relative), but I decided that was also beyond the scope of this PR (I'd prefer to focus this PR on dynamic linker paths only).

> Thanks for the PR!

Thanks for being open to this change. Let's see if we can get it into some mergeable state.

gerlero commented:
> To make sure we're on the same page about all the consequences of this pull request: can you show an example of the output of `otool -L` for a dylib before and after this change?

Sure, here's an output for `open-mpi` for example (```otool -L /opt/homebrew/opt/open-mpi/lib/libmpi.dylib```):

Before:
```
/opt/homebrew/opt/open-mpi/lib/libmpi.dylib:
	/opt/homebrew/opt/open-mpi/lib/libmpi.40.dylib (compatibility version 71.0.0, current version 71.5.0)
	/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-rte.40.dylib (compatibility version 71.0.0, current version 71.3.0)
	/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-pal.40.dylib (compatibility version 71.0.0, current version 71.3.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/opt/homebrew/opt/hwloc/lib/libhwloc.15.dylib (compatibility version 22.0.0, current version 22.2.0)
	/opt/homebrew/opt/libevent/lib/libevent_core-2.1.7.dylib (compatibility version 8.0.0, current version 8.1.0)
	/opt/homebrew/opt/libevent/lib/libevent_pthreads-2.1.7.dylib (compatibility version 8.0.0, current version 8.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

After:
```
/opt/homebrew/opt/open-mpi/lib/libmpi.dylib:
	/opt/homebrew/opt/open-mpi/lib/libmpi.40.dylib (compatibility version 71.0.0, current version 71.5.0)
	@loader_path/libopen-rte.40.dylib (compatibility version 71.0.0, current version 71.3.0)
	@loader_path/libopen-pal.40.dylib (compatibility version 71.0.0, current version 71.3.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	@loader_path/../../../../opt/hwloc/lib/libhwloc.15.dylib (compatibility version 22.0.0, current version 22.2.0)
	@loader_path/../../../../opt/libevent/lib/libevent_core-2.1.7.dylib (compatibility version 8.0.0, current version 8.1.0)
	@loader_path/../../../../opt/libevent/lib/libevent_pthreads-2.1.7.dylib (compatibility version 8.0.0, current version 8.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

gerlero commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -22,17 +22,17 @@ def relocate_dynamic_linkage(relocation)
     mach_o_files.each do |file|
       file.ensure_writable do
         if file.dylib?
-          id = relocated_name_for(file.dylib_id, relocation)
+          id = relocated_name_for(file, file.dylib_id, relocation)
           change_dylib_id(id, file)
         end
```
Thanks for the review and find. You're right.

For some reason I haven't been able to track down, this change in the `id` is not actually being made (`change_dylib_id` is called but doesn't seem to do anything). But, if it the change did happen, then the `id` of the library would be set to a relative path, which would present an incorrect path to other software that wants to use this dylib.

My intent with this PR is not to touch the `id`s at all. I don't like the fact the `id`s are also absolute paths, as that's another barrier to useful relocation, but I couldn't think of an easy fix for this (I believe there's no simple way to define a relative `id` that would immediately work for other software linking against that lib?).

For now, I'll find a way to fully remove the relative-path logic from `change_dylib_id` path.

gerlero replied:
Should be fixed now

gerlero commented:
@carlocab I've updated the `LinkageChecker` so it now expands install names with `@loader_path` instead of skipping them.

I've also added expansion of `@executable_path` when the file in question is an executable, as in that case it will always be a synonym for `@loader_path` (unless one can load another executable as a library? I don't believe that to be possible).

gerlero commented:
@carlocab and @MikeMcQuaid Thanks again!

I've factored out the new name substitutions into a `loader_name_for` function (can be bikeshedded), which only gets called when dealing with `install_name`s and `rpath`s (and never for `id`s), and only at two clear points: when fixing paths of a build from source (`fix_dynamic_linkage`), and after pouring a bottle (`relocate_dynamic_linkage`).

Following @carlocab's suggestion, I've also updated `linkage_checker.rb` so that it it is now able to check these relative paths to ensure they are not broken. I see that this is tested by the "test default formula" jobs, which are passing (BTW, the "cask audit" job failure looks like just some server being down). This part gives me a little more confidence that this change won't immediately break other people's installs.

I'll wait for more comments/reviews/guidance before making any more changes. 

P.S. I think we could also add a test to prevent any accidental regressions, but I'm not sure where that would go.

carlocab commented:
I'm currently neck-deep in Homebrew/homebrew-core#134251, so I haven't had a chance to take a proper look here. Will come back to it next week.

carlocab commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -168,21 +168,29 @@ def check_dylibs(rebuild_cache:)
 
-        if dylib.start_with? "@"
+        if dylib.start_with?("@rpath") ||
+           (dylib.start_with?("@executable_path") && !pathname.binary_executable?)
```
This case is a little worrying. While it is technically possible to do this, it means that the linked library (`dylib`) that is loaded depends on the location of the executable that loads `file`. I feel like we should be throwing at least a warning for these with `brew linkage --test --strict`. (But this isn't needed for your PR -- unless you're feeling generous.)

gerlero replied:
Not gonna say it'd wrong with no exceptions (I believe doing that would be fine e.g. within an app bundle that doesn't expect external binaries to link against its libraries), but I don't really expect to see such a thing in Homebrew.

I can agree with a warning + a failure if `--strict` (and I guess I'm feeling generous! 🙂). Done!

carlocab commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -28,12 +28,14 @@ def relocate_dynamic_linkage(relocation)
         each_linkage_for(file, :dynamically_linked_libraries) do |old_name|
           new_name = relocated_name_for(old_name, relocation)
-          change_install_name(old_name, new_name, file) if new_name
+          new_name = loader_name_for(file, new_name)
```
Can we also avoid recycling the `new_name` variable? I feel like it makes understanding what's going on a bit more confusing than it should be. (But feel free to leave it as is if you have a strong opinion here.)

gerlero replied:
No strong opinion. I've changed it (but let me know if you think of a better name or names).

carlocab commented on Library/Homebrew/linkage_checker.rb:
Could you also show the output of `brew linkage` and `brew linkage --reverse` before and after your changes here?

It would also be good to verify that `brew linkage --test` fails if the wanted library dependency has gone missing (e.g. now has a different install name, is no longer on disk, etc).

gerlero replied:
> Could you also show the output of `brew linkage` and `brew linkage --reverse` before and after your changes here?

**Before**

```sh
brew linkage open-mpi
```

```        
System libraries:
  /usr/lib/libSystem.B.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /opt/homebrew/opt/gcc/lib/gcc/current/libgfortran.5.dylib (gcc)
  /opt/homebrew/opt/gcc/lib/gcc/current/libquadmath.0.dylib (gcc)
  /opt/homebrew/opt/hwloc/lib/libhwloc.15.dylib (hwloc)
  /opt/homebrew/opt/libevent/lib/libevent_core-2.1.7.dylib (libevent)
  /opt/homebrew/opt/libevent/lib/libevent_pthreads-2.1.7.dylib (libevent)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_dstore.1.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_monitoring.50.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_ompio.41.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_sm.40.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmpi.40.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmpi_mpifh.40.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-pal.40.dylib (open-mpi)
  /opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-rte.40.dylib (open-mpi)
```

```sh
brew linkage --reverse open-mpi
```
<details>
/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_dstore.1.dylib
  lib/pmix/mca_gds_ds12.so
  lib/pmix/mca_gds_ds21.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_monitoring.50.dylib
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_pml_monitoring.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_ompio.41.dylib
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmca_common_sm.40.dylib
  lib/openmpi/mca_coll_sm.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmpi.40.dylib
  bin/ompi_info
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempif08.40.dylib

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-pal.40.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/opt/homebrew/Cellar/open-mpi/4.1.5/lib/libopen-rte.40.dylib
  bin/ompi_info
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/opt/homebrew/opt/gcc/lib/gcc/current/libgfortran.5.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib

/opt/homebrew/opt/gcc/lib/gcc/current/libquadmath.0.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib

/opt/homebrew/opt/hwloc/lib/libhwloc.15.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmca_common_monitoring.50.dylib
  lib/libmca_common_ompio.41.dylib
  lib/libmca_common_sm.40.dylib
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libompitrace.40.dylib
  lib/libopen-pal.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/libompi_dbg_msgq.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_tcp.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_adapt.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_han.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_io_romio321.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pml_ob1.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/opt/homebrew/opt/libevent/lib/libevent_core-2.1.7.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmca_common_monitoring.50.dylib
  lib/libmca_common_ompio.41.dylib
  lib/libmca_common_sm.40.dylib
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libompitrace.40.dylib
  lib/libopen-pal.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/libompi_dbg_msgq.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_tcp.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_adapt.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_han.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_io_romio321.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pml_ob1.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/opt/homebrew/opt/libevent/lib/libevent_pthreads-2.1.7.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmca_common_monitoring.50.dylib
  lib/libmca_common_ompio.41.dylib
  lib/libmca_common_sm.40.dylib
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libompitrace.40.dylib
  lib/libopen-pal.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/libompi_dbg_msgq.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_tcp.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_adapt.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_han.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_io_romio321.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pml_ob1.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so

/usr/lib/libSystem.B.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmca_common_dstore.1.dylib
  lib/libmca_common_monitoring.50.dylib
  lib/libmca_common_ompio.41.dylib
  lib/libmca_common_sm.40.dylib
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libompitrace.40.dylib
  lib/libopen-pal.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/libompi_dbg_msgq.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_tcp.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_adapt.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_han.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_io_romio321.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pml_ob1.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so
  lib/pmix/mca_bfrops_v12.so
  lib/pmix/mca_bfrops_v20.so
  lib/pmix/mca_bfrops_v21.so
  lib/pmix/mca_bfrops_v3.so
  lib/pmix/mca_gds_ds12.so
  lib/pmix/mca_gds_ds21.so
  lib/pmix/mca_gds_hash.so
  lib/pmix/mca_pcompress_zlib.so
  lib/pmix/mca_plog_default.so
  lib/pmix/mca_plog_stdfd.so
  lib/pmix/mca_plog_syslog.so
  lib/pmix/mca_preg_compress.so
  lib/pmix/mca_preg_native.so
  lib/pmix/mca_psec_native.so
  lib/pmix/mca_psec_none.so
  lib/pmix/mca_psensor_file.so
  lib/pmix/mca_psensor_heartbeat.so
  lib/pmix/mca_pshmem_mmap.so
  lib/pmix/mca_psquash_flex128.so
  lib/pmix/mca_psquash_native.so
  lib/pmix/mca_ptl_tcp.so
  lib/pmix/mca_ptl_usock.so

/usr/lib/libz.1.dylib
  bin/ompi_info
  bin/opal_wrapper
  bin/orte-clean
  bin/orte-info
  bin/orte-server
  bin/orted
  bin/orterun
  lib/libmca_common_dstore.1.dylib
  lib/libmca_common_monitoring.50.dylib
  lib/libmca_common_ompio.41.dylib
  lib/libmca_common_sm.40.dylib
  lib/libmpi.40.dylib
  lib/libmpi_mpifh.40.dylib
  lib/libmpi_usempi_ignore_tkr.40.dylib
  lib/libmpi_usempif08.40.dylib
  lib/libompitrace.40.dylib
  lib/libopen-pal.40.dylib
  lib/libopen-rte.40.dylib
  lib/ompi_monitoring_prof.so
  lib/openmpi/libompi_dbg_msgq.so
  lib/openmpi/mca_allocator_basic.so
  lib/openmpi/mca_allocator_bucket.so
  lib/openmpi/mca_bml_r2.so
  lib/openmpi/mca_btl_self.so
  lib/openmpi/mca_btl_sm.so
  lib/openmpi/mca_btl_tcp.so
  lib/openmpi/mca_btl_vader.so
  lib/openmpi/mca_coll_adapt.so
  lib/openmpi/mca_coll_basic.so
  lib/openmpi/mca_coll_han.so
  lib/openmpi/mca_coll_inter.so
  lib/openmpi/mca_coll_libnbc.so
  lib/openmpi/mca_coll_monitoring.so
  lib/openmpi/mca_coll_self.so
  lib/openmpi/mca_coll_sm.so
  lib/openmpi/mca_coll_sync.so
  lib/openmpi/mca_coll_tuned.so
  lib/openmpi/mca_compress_bzip.so
  lib/openmpi/mca_compress_gzip.so
  lib/openmpi/mca_crs_none.so
  lib/openmpi/mca_errmgr_default_app.so
  lib/openmpi/mca_errmgr_default_hnp.so
  lib/openmpi/mca_errmgr_default_orted.so
  lib/openmpi/mca_errmgr_default_tool.so
  lib/openmpi/mca_ess_env.so
  lib/openmpi/mca_ess_hnp.so
  lib/openmpi/mca_ess_pmi.so
  lib/openmpi/mca_ess_singleton.so
  lib/openmpi/mca_ess_slurm.so
  lib/openmpi/mca_ess_tool.so
  lib/openmpi/mca_fbtl_posix.so
  lib/openmpi/mca_fcoll_dynamic.so
  lib/openmpi/mca_fcoll_dynamic_gen2.so
  lib/openmpi/mca_fcoll_individual.so
  lib/openmpi/mca_fcoll_two_phase.so
  lib/openmpi/mca_fcoll_vulcan.so
  lib/openmpi/mca_filem_raw.so
  lib/openmpi/mca_fs_ufs.so
  lib/openmpi/mca_grpcomm_direct.so
  lib/openmpi/mca_io_ompio.so
  lib/openmpi/mca_io_romio321.so
  lib/openmpi/mca_iof_hnp.so
  lib/openmpi/mca_iof_orted.so
  lib/openmpi/mca_iof_tool.so
  lib/openmpi/mca_mpool_hugepage.so
  lib/openmpi/mca_odls_default.so
  lib/openmpi/mca_odls_pspawn.so
  lib/openmpi/mca_oob_tcp.so
  lib/openmpi/mca_osc_monitoring.so
  lib/openmpi/mca_osc_pt2pt.so
  lib/openmpi/mca_osc_rdma.so
  lib/openmpi/mca_osc_sm.so
  lib/openmpi/mca_patcher_overwrite.so
  lib/openmpi/mca_plm_isolated.so
  lib/openmpi/mca_plm_rsh.so
  lib/openmpi/mca_plm_slurm.so
  lib/openmpi/mca_pmix_flux.so
  lib/openmpi/mca_pmix_isolated.so
  lib/openmpi/mca_pmix_pmix3x.so
  lib/openmpi/mca_pml_cm.so
  lib/openmpi/mca_pml_monitoring.so
  lib/openmpi/mca_pml_ob1.so
  lib/openmpi/mca_pstat_test.so
  lib/openmpi/mca_ras_gridengine.so
  lib/openmpi/mca_ras_simulator.so
  lib/openmpi/mca_ras_slurm.so
  lib/openmpi/mca_rcache_grdma.so
  lib/openmpi/mca_reachable_weighted.so
  lib/openmpi/mca_regx_fwd.so
  lib/openmpi/mca_regx_naive.so
  lib/openmpi/mca_regx_reverse.so
  lib/openmpi/mca_rmaps_mindist.so
  lib/openmpi/mca_rmaps_ppr.so
  lib/openmpi/mca_rmaps_rank_file.so
  lib/openmpi/mca_rmaps_resilient.so
  lib/openmpi/mca_rmaps_round_robin.so
  lib/openmpi/mca_rmaps_seq.so
  lib/openmpi/mca_rml_oob.so
  lib/openmpi/mca_routed_binomial.so
  lib/openmpi/mca_routed_direct.so
  lib/openmpi/mca_routed_radix.so
  lib/openmpi/mca_rtc_hwloc.so
  lib/openmpi/mca_schizo_flux.so
  lib/openmpi/mca_schizo_jsm.so
  lib/openmpi/mca_schizo_ompi.so
  lib/openmpi/mca_schizo_orte.so
  lib/openmpi/mca_schizo_slurm.so
  lib/openmpi/mca_sharedfp_individual.so
  lib/openmpi/mca_sharedfp_lockedfile.so
  lib/openmpi/mca_sharedfp_sm.so
  lib/openmpi/mca_shmem_mmap.so
  lib/openmpi/mca_shmem_posix.so
  lib/openmpi/mca_shmem_sysv.so
  lib/openmpi/mca_state_app.so
  lib/openmpi/mca_state_hnp.so
  lib/openmpi/mca_state_novm.so
  lib/openmpi/mca_state_orted.so
  lib/openmpi/mca_state_tool.so
  lib/openmpi/mca_topo_basic.so
  lib/openmpi/mca_topo_treematch.so
  lib/openmpi/mca_vprotocol_pessimist.so
  lib/pmix/mca_bfrops_v12.so
  lib/pmix/mca_bfrops_v20.so
  lib/pmix/mca_bfrops_v21.so
  lib/pmix/mca_bfrops_v3.so
  lib/pmix/mca_gds_ds12.so
  lib/pmix/mca_gds_ds21.so
  lib/pmix/mca_gds_hash.so
  lib/pmix/mca_pcompress_zlib.so
  lib/pmix/mca_plog_default.so
  lib/pmix/mca_plog_stdfd.so
  lib/pmix/mca_plog_syslog.so
  lib/pmix/mca_preg_compress.so
  lib/pmix/mca_preg_native.so
  lib/pmix/mca_psec_native.so
  lib/pmix/mca_psec_none.so
  lib/pmix/mca_psensor_file.so
  lib/pmix/mca_psensor_heartbeat.so
  lib/pmix/mca_pshmem_mmap.so
  lib/pmix/mca_psquash_flex128.so
  lib/pmix/mca_psquash_native.so
  lib/pmix/mca_ptl_tcp.so
  lib/pmix/mca_ptl_usock.so

</details>

gerlero replied:
**After**
```sh
brew linkage open-mpi
```

_100% exact same output as **before**_

```sh
brew linkage --reverse open-mpi
```

_100% exact same output as **before**_

gerlero replied:
> It would also be good to verify that brew linkage --test fails if the wanted library dependency has gone missing (e.g. now has a different install name, is no longer on disk, etc).

```sh
brew linkage --test open-mpi
```

_OK (no output)_

**Let's purposely break some stuff**

```sh
install_name_tool -change @loader_path/../../../../opt/hwloc/lib/libhwloc.15.dylib @loader_path/../../../../opt/hwloc/lib/libFAIL.15.dylib /opt/homebrew/opt/open-mpi/lib/libmpi.dylib
install_name_tool -change @loader_path/libopen-rte.40.dylib @executable_path/libopen-rte.40.dylib /opt/homebrew/opt/open-mpi/lib/libmpi.dylib
```

_Output (code signature warnings)_
```
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /opt/homebrew/opt/open-mpi/lib/libmpi.dylib
/Library/Developer/CommandLineTools/usr/bin/install_name_tool: warning: changes being made to the file will invalidate the code signature in: /opt/homebrew/opt/open-mpi/lib/libmpi.dylib
```

**Linkage tests** (after clearing the cache)

Non-strict:

```sh
brew linkage --test open-mpi
```
```
Broken dependencies:
  /opt/homebrew/opt/hwloc/lib/libFAIL.15.dylib (hwloc)
```

Strict:

```sh
brew linkage --test --strict open-mpi
```

```
Broken dependencies:
  /opt/homebrew/opt/hwloc/lib/libFAIL.15.dylib (hwloc)
@executable_path references in libraries:
  @executable_path/libopen-rte.40.dylib
```

gerlero commented on Library/Homebrew/linkage_checker.rb:
```diff
@@ -168,21 +168,29 @@ def check_dylibs(rebuild_cache:)
     keg_files_dylibs.each do |file, dylibs|
       file_has_any_rpath_dylibs = T.let(false, T::Boolean)
       dylibs.each do |dylib|
+        pathname = Pathname(file)
+        # Resolve some variable install names.
+        expanded_path = pathname.dirname
+        dylib.sub!("@loader_path", expanded_path.to_s)
+        dylib.sub!("@executable_path", expanded_path.to_s) if pathname.binary_executable?
+        # Remove redundant dots from the path.
+        dylib = Pathname(dylib).cleanpath.to_s
```
I believe moving it further up the call stack to `#dynamically_linked_libraries` would make it so that the call to `#each_linkage_for` in `#fix_dynamic_linkage` will receive the resolved paths for any relative linkages and attempt to needlessly fix it again (using an inexistent name in the process)?

carlocab replied:
We could give `#dynamically_linked_libraries` an optional kwarg, e.g. `resolve_variable_references`, so that calling
```
file.dynamically_linked_libraries
```
will return resolved paths, while
```
file.dynamically_linked_libraries(resolve_variable_references: false)
```
will not. You can use the latter in `#fix_dynamic_linkage`.

gerlero replied:
Not opposed to it. I like the idea of moving this logic to the `MachOShim` class and controlling it via a `resolve_variable_references` kwarg (that defaults to `true`), although I think a name like that could mean the method may be expected to also resolve `@rpath`-based install names (which it currently won't and isn't needed to check the changes in this PR; although that could be added in a future improvement to `LinkageChecker`).

If you're okay with that (I'm fine with it), I can make this change.

gerlero commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -28,12 +28,14 @@ def relocate_dynamic_linkage(relocation)
           new_name = relocated_name_for(old_name, relocation)
-          change_rpath(old_name, new_name, file) if new_name
+          new_name = loader_name_for(file, new_name)
+          change_rpath(old_name, new_name, file) if new_name != old_name
```
You're right. I had the wrong idea of what methods were called in each of the scenario. Looks like no changes to `#relocate_dynamic_linkage` are actually needed, for two reasons in fact:

1. `#fix_dynamic_linkage` is called anyway even when pouring from bottles (I didn't realize that). This makes it so that even "old" bottles get the relative names on install.

2. Bottles built after this change already have relative `install_name`s and `rpath`s, making much of the work done by `#relocate_dynamic_linkage` eventually unnecessary, as you mention.

**I've removed all changes from `#relocate_dynamic_linkage`, keeping just the changes to `#fix_dynamic_linkage`.**

> In fact, the rewriting absolute references to `$HOMEBREW_PREFIX` should make a significant chunk of `#relocate_dynamic_linkage` unnecessary. What do you think?

Indeed, after every bottle is rebuilt with this change, these lines should no longer be needed:

https://github.com/Homebrew/brew/blob/c5cbe6406b8b9d040c5d1850b5f041af4ddef83c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L29-L37

(Some kind of deprecation cycle might be a good idea to let maintainers of third-party taps know that they should rebuild their bottles too)

gerlero replied:
BTW, be aware that this PR assumes that `HOMEBREW_CELLAR` is a subdirectory of `HOMEBREW_PREFIX` (the actual name shouldn't matter). 

I mention it because I see that the relocation code deals `HOMEBREW_CELLAR` and `HOMEBREW_PREFIX` somewhat separately. I haven't seen a Homebrew install where `HOMEBREW_CELLAR != HOMEBREW_PREFIX/Cellar`, but if such an install exists, it will be broken by this PR.

carlocab commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -45,27 +45,40 @@ def fix_dynamic_linkage
   end
 
+  def loader_name_for(file, target)
+    # Use @loader_path-relative install names for other Homebrew-installed binaries.
+    if target.start_with?(HOMEBREW_PREFIX)
+      "@loader_path/#{Pathname(target).relative_path_from(file.dirname)}"
+    else
+      target
+    end
```
I wonder if we could/should gate this behind an environment variable (e.g. `HOMEBREW_USE_RELATIVE_INSTALL_NAMES`) first?

We can set it in Homebrew/core CI for testing, and then remove the environment variable check when we're ready. But at least the env check will allow us to quickly turn it off if we discover any serious problems with it.

It'll also help us fine-tune the timing of when this change will take effect in Homebrew/core CI. I suspect we'll actually want to make sure that the `linkage_checker` changes lands in a release tag first before building bottles with this enabled.

gerlero replied:
An alternative I can think of now (could be combined with an env var) that would cause this PR to take effect gradually would be to add an extra check in the code so that these new lines quoted here aren't run when pouring from a bottle (as of now, the code in the quoted snippet is called from `#fix_dynamic_linkage`, which can get called regardless of if installing from source or a bottle).

Would need some more changes in the code to achieve this (guess the simplest way would be to add a flag to `#fix_dynamic_linkage` that tells it if it's pouring from a bottle or not), but it would make it so that only newly built bottles and installs from source are affected by the change, containing the extent of immediate consequences of any negative impact from this PR. Also, as discussed above, bottles built after this change will already have relative install names in their binaries and don't really need to go through this rewrite step at all.

MikeMcQuaid replied:
> I wonder if we could/should gate this behind an environment variable (e.g. `HOMEBREW_USE_RELATIVE_INSTALL_NAMES`) first?

Agreed. This is a blocker for me, I think, sorry!

Ideally all the same behaviour would remain unless one or more environment variables are set. I am fine with there being one variable to handle bottle creation and another for bottle pouring or one variable for both; whatever consensus @carlocab @gerlero agree with works for me.

gerlero replied:
@MikeMcQuaid Your call. I'm not sure then whether to (a) gate relative install names when pouring from bottles with the same env variable, (b) control what to do with bottle installs with a different environment variable, or (c) not do anything new at all when pouring from bottles.

(b) would be the most flexible but I don't that flexibility is really needed... (c) is the ultimate goal of this PR; but if the idea of introducing this behind an env var is for the feature to be tested more thoroughly, then (a) should create more opportunities for that.

@carlocab do you have a particular stance on this?

MikeMcQuaid replied:
@gerlero My take is that any functionality that changes with creating or pouring bottles should be behind 1 or 2 environment variables for now. I don't have any preference whether it's 1 or 2: you decide. If @carlocab replies today with a preference: do what he says otherwise you just do whatever you want (which sounds like 1 which is fine/good).

Thanks again, sorry for all the back and forth!

gerlero replied:
> Suggestion: make `#loader_name_for` be no-op (i.e. always take the `else` branch) unless `HOMEBREW_RELOCATABLE_INSTALL_NAMES` is set.

Great. I can do that.

> If my understanding of how this all works is correct: the changes here should only matter when creating bottles, and not when pouring them, so there is no need for an extra environment variable that controls behaviour when pouring bottles.

Actually, the changes here as of now cover both, as `#fix_dynamic_linkage` is called when pouring bottles too:

https://github.com/Homebrew/brew/blob/a34d46e0938e2a830b8466729ffb2fb60de9a80e/Library/Homebrew/formula_installer.rb#L786

I apologize for this confusion; I got this part wrong when I first opened the PR (incorrectly assumed that `#fix_dynamic_linkage` was only invoked for builds from source), so this might be the source of the confusion.

As it is now, **bottles built without relative names will be changed to use relative names on install**. If this is undesired, then the new logic will need to be tweaked a bit to be able to detect the type of install and disable the behavior it it's a bottle pour.

> The `linkage_checker` changes are an improvement regardless of whether we adopt your `keg_relocate` changes or not, so these should be shipped to users as soon as it's sensible to.

I agree (that part could've even been its own PR). I'll move the new logic to `MachOShim` and then we should be done with that for now.

> When this happens, we can try setting `HOMEBREW_RELOCATABLE_INSTALL_NAMES` in Homebrew/core CI and see try these changes out. Once we're happy, we can remove the environment variable check.

Sounds good.

I'll let you know then I'm done with the new changes.

carlocab replied:
> Actually, the changes here as of now cover both, as `#fix_dynamic_linkage` is called when pouring bottles too:

Yes, I noticed that too. My point is that after bottles are built with these changes: `fix_dynamic_linkage` should have no effect when pouring bottles.

gerlero replied:
@MikeMcQuaid @carlocab Thanks again for all your comments. I've pushed the changes (added the `HOMEBREW_RELOCATABLE_INSTALL_NAMES` env var check and moved the expansion of relative install names to `MachOShim`).

carlocab reviewed (approved):
Happy with how this looks; thanks for your work here!

There are some timing questions about merging this that would be good to think about, though. See comments below + CC @MikeMcQuaid @Bo98 for thoughts.

carlocab commented:
Another thought that occurs to me, that would be useful for your use case, but would likely have to stay permanently gated behind an environment variable (e.g. `HOMEBREW_USE_RPATH_PREFIXES`):

If this environment variable is set, `brew` will rename all dylib IDs so that they are `@rpath`-prefixed, e.g., `@rpath/libbasename.1.2.3.dylib`. This should be implemented alongside #15354, since the desired behaviour is similar (except one is driven by a formula definition, the other by an environment variable).

gerlero commented:
> Another thought that occurs to me, that would be useful for your use case, but would likely have to stay permanently gated behind an environment variable (e.g. `HOMEBREW_USE_RPATH_PREFIXES`):
> 
> If this environment variable is set, `brew` will rename all dylib IDs so that they are `@rpath`-prefixed, e.g., `@rpath/libbasename.1.2.3.dylib`. This should be implemented alongside #15354, since the desired behaviour is similar (except one is driven by a formula definition, the other by an environment variable).

Thanks! Interesting. I did play around with something similar to that just before opening this PR: I actually chose `$HOMEBREW_PREFIX` as the anchor point for `rpath` and then replaced every `install_name` and `id` to be defined relative to `@rpath`. In that case, you can easily link anything as long as there's an `rpath` that points to the Homebrew install (via a compiler/linker flag or by adding an `rpath` entry afterwards—you can use a relative `rpath` for maximum relocatability!).

You're right that my use case could benefit from something like this, even if it needs to be gated behind an env var as it would break stuff for regular users (I still wonder whether a change like this can be made to work in a transparent way for regular usage too...)

MikeMcQuaid reviewed (approved):
Good with me when @carlocab is happy!

carlocab reviewed (approved):
Nice work. Thanks very much!

carlocab commented:
Let's ship this so we can get to testing it soon-ish.

fxcoudert commented:
This broke julia: https://github.com/Homebrew/homebrew-core/pull/135398#issuecomment-1643041541 by altering the order of rpath's (see details in the linked comment and the one above).

carlocab commented:
I don't think it's this change, because this shouldn't affect rpaths unless `HOMEBREW_RELOCATABLE_INSTALL_NAMES` is set.

https://github.com/Homebrew/brew/commit/27032e002fb51fc5554145cef1a54fba2b32505f sounds more likely.

cho-m commented:
Didn't https://github.com/Homebrew/brew/commit/27032e002fb51fc5554145cef1a54fba2b32505f go in 3 days ago?

RPATH issue has been happening for a week now, e.g. https://github.com/Homebrew/homebrew-core/pull/136273

cho-m commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -45,27 +45,40 @@ def fix_dynamic_linkage
-                  (file.rpaths.count(bad_name) == 1)
-
-          delete_rpath(bad_name, file)
+          if rooted_in_build_directory?(bad_name) || (file.rpaths.count(bad_name) > 1)
+            delete_rpath(bad_name, file)
```
`file.rpaths` has resolved `@loader_path/` leading to `julia` issue. Printing out the data:
```
["/usr/local/Cellar/julia/1.9.2/lib/julia",
 "/usr/local/opt/curl/lib",
 "/usr/local/opt/mbedtls@2/lib",
 "/usr/local/opt/openblas/lib",
 "/usr/local/opt/gcc/lib/gcc/current",
 "/usr/local/lib",
 "/usr/lib",
 "/usr/local/Cellar/julia/1.9.2/lib/julia"]
==> Deleting rpath /usr/local/Cellar/julia/1.9.2/lib/julia in /usr/local/Cellar/julia/1.9.2/lib/julia/libcholmod.3.0.14.dylib
```

gerlero replied:
> file.rpaths has resolved @loader_path/ leading to julia issue where earlier RPATH is deleted changing which dylib gets loaded

@cho-m Yeah, that's bad and is caused by the changes in this PR (specifically, https://github.com/Homebrew/brew/pull/15571/commits/15a0c7fd7df2cae55e3b702a04e7bbf1ad401f27). 

Changing the ``files.rpaths.count(bad_name)`` call to ``files.rpaths(resolve_variable_references: false).count(bad_name)`` should go back to the old behavior, which I think is the correct behavior here (@carlocab please correct me if I'm wrong).

carlocab replied:
Yes, Julia should be fixed now.

gerlero replied:
Right, fix is in #15722. Thanks!
